src/modules/resource/discover.c

```diff
@@ -0,0 +1,289 @@
+/************************************************************\
+ * Copyright 2020 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* discover.c - dynamic resource discovery
+ *
+ * Populate resource.hwloc and provide access to resource.hwloc.by_rank
+ * on demand.
+ *
+ * At initialization, the eventlog is replayed.  If events (or lack thereof)
+ * indicate that resource.hwloc is not already populated, run
+ * "flux hwloc reload".  Since that requires all ranks to be up, it is not run
+ * until the monitor subsystem says they are.
+ *
+ * In addition, once resource.hwloc is populated, the 'resource.hwloc.by_rank'
+ * object is looked up from the KVS and made available via discover_get().
+
+ * For testing, resource discovery may be defeated by populating
+ * resource.hwloc.by_rank with dummy resources and posting a fake
+ * hwloc-discover-finish event to 'resource.eventlog' before loading
+ * this module.
+ *
+ * Caveats:
+ * - resource.by_rank is a stand-in for future Flux concrete resource object
+ * - no support for statically configured resources yet
+ * - no support for obtaining resources from enclosing instance
+ * - all ranks have to be online before 'flux hwloc reload' can run
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <unistd.h>
+#include <jansson.h>
+#include <flux/core.h>
+#include <assert.h>
+
+#include "src/common/libidset/idset.h"
+#include "src/common/libeventlog/eventlog.h"
+
+#include "resource.h"
+#include "reslog.h"
+#include "discover.h"
+#include "monitor.h"
+
+struct discover {
+    struct resource_ctx *ctx;
+    flux_subprocess_t *p;
+    bool ready;             // all broker ranks are online
+    bool loaded;            // resource.hwloc is populated
+    flux_future_t *f;
+};
+
+static const char *auxkey = "flux::discover";
+
+const json_t *discover_get (struct discover *discover)
+{
+    json_t *by_rank;
+
+    if (!discover->f)
+        return NULL;
+    if (flux_kvs_lookup_get_unpack (discover->f, "o", &by_rank) < 0) {
+        flux_log_error (discover->ctx->h, "hwloc.by_rank");
+        return NULL;
+    }
+    return by_rank;
+}
+
+static void lookup_hwloc_continuation (flux_future_t *f, void *arg)
+{
+    struct discover *discover = arg;
+    struct resource_ctx *ctx = discover->ctx;
+
+    if (flux_future_get (f, NULL) < 0)
+        flux_log_error (ctx->h, "hwloc.by_rank");
+    /* discover->f remains valid so by_rank can be accessed later */
+}
+
+static int lookup_hwloc (struct discover *discover)
+{
+    struct resource_ctx *ctx = discover->ctx;
+
+    assert (discover->f == NULL);
+    assert (discover->loaded);
```

Comment:
Do we need this assert check? You probably have a good reason for it but since `assert()` will kill the entire flux instance, this caught my eye. 

By the way, in general when do we call terminating assert()? Like when it smells like a memory corruption?

Reply:
The intent was to document an invariant, but you're probably right, that's not a great idea in module/broker code.